File: banking_backend/services/exchange_rate.py
"""
Exchange Rate Service - Currency Agnostic
Fetches live exchange rates and converts between ANY currencies
NO HARDCODED CURRENCIES!
"""

import requests
from typing import Dict, Optional
from datetime import datetime, timedelta


class ExchangeRateService:
    """
    Currency-agnostic exchange rate service.
    Works with ANY base currency, not just NGN!
    """
    
    def __init__(self, default_base: str = 'USD'):
        """
        Initialize exchange rate service.
        
        Args:
            default_base: Default base currency (USD recommended for global system)
        """
        self.default_base = default_base
        self.cache: Dict[str, Dict] = {}  # Cache per base currency
        self.cache_duration = timedelta(hours=1)
        self.cache_expiry: Dict[str, datetime] = {}
    
    def get_rates(self, base_currency: str = None) -> Dict[str, float]:
        """
        Get exchange rates FROM a base currency to all other currencies.
        
        Args:
            base_currency: Base currency (e.g., 'USD', 'EUR', 'NGN')
                          If None, uses default_base
        
        Returns:
            Dict of currency codes to exchange rates
            Example: {'EUR': 0.93, 'GBP': 0.79, ...}
        """
        base = (base_currency or self.default_base).upper()
        
        # Check cache
        if base in self.cache and base in self.cache_expiry:
            if datetime.now() < self.cache_expiry[base]:
                return self.cache[base]
        
        # Fetch fresh rates
        try:
            url = f"https://api.exchangerate-api.com/v4/latest/{base}"
            response = requests.get(url, timeout=5)
            response.raise_for_status()
            
            data = response.json()
            rates = data.get('rates', {})
            
            # Update cache
            self.cache[base] = rates
            self.cache_expiry[base] = datetime.now() + self.cache_duration
            
            return rates
            
        except requests.RequestException as e:
            print(f"Failed to fetch exchange rates for {base}: {e}")
            
            # Return fallback rates if available
            if base in self.cache:
                return self.cache[base]
            
            # Last resort fallback
            return self._get_fallback_rates(base)
    
    def _get_fallback_rates(self, base: str) -> Dict[str, float]:
        """
        Fallback exchange rates when API is unavailable.
        Uses approximate rates (updated periodically).
        """
        # Common fallback rates (as of Dec 2024)
        if base == 'USD':
            return {
                'USD': 1.0,
                'EUR': 0.93,
                'GBP': 0.79,
                'JPY': 150.0,
                'NGN': 1500.0,
                'GHS': 12.0,
                'ZAR': 18.5,
                'CAD': 1.35,
                'AUD': 1.52,
                'CHF': 0.88,
                'CNY': 7.24,
                'INR': 83.0,
            }
        elif base == 'NGN':
            return {
                'NGN': 1.0,
                'USD': 1/1500.0,
                'EUR': 1/1600.0,
                'GBP': 1/1900.0,
                'JPY': 0.10,
            }
        else:
            # Return USD-based rates if we don't have specific fallback
            return {'USD': 1.0}
    
    def convert(
        self,
        amount: float,
        from_currency: str,
        to_currency: str = None
    ) -> Dict:
        """
        Convert amount from one currency to another.
        
        Args:
            amount: Amount to convert
            from_currency: Source currency code
            to_currency: Target currency code (defaults to default_base)
        
        Returns:
            Dict with converted_amount, from_currency, to_currency, rate
        """
        to_curr = (to_currency or self.default_base).upper()
        from_curr = from_currency.upper()
        
        # Same currency = no conversion needed
        if from_curr == to_curr:
            return {
                'amount': amount,
                'converted_amount': amount,
                'from_currency': from_curr,
                'to_currency': to_curr,
                'rate': 1.0
            }
        
        # Get rates FROM source currency
        rates = self.get_rates(from_curr)
        
        # Get rate to target currency
        rate = rates.get(to_curr, 1.0)
        converted = amount * rate
        
        return {
            'amount': amount,
            'converted_amount': round(converted, 2),
            'from_currency': from_curr,
            'to_currency': to_curr,
            'rate': round(rate, 6)
        }
    
    def get_rate(self, from_currency: str, to_currency: str = None) -> float:
        """
        Get exchange rate between two currencies.
        
        Args:
            from_currency: Source currency
            to_currency: Target currency (defaults to default_base)
        
        Returns:
            Exchange rate as float
        """
        to_curr = (to_currency or self.default_base).upper()
        from_curr = from_currency.upper()
        
        if from_curr == to_curr:
            return 1.0
        
        rates = self.get_rates(from_curr)
        return rates.get(to_curr, 1.0)
    
    def clear_cache(self, base_currency: str = None):
        """Clear cache for specific or all currencies."""
        if base_currency:
            if base_currency in self.cache:
                del self.cache[base_currency]
            if base_currency in self.cache_expiry:
                del self.cache_expiry[base_currency]
        else:
            self.cache.clear()
            self.cache_expiry.clear()


# Singleton instance with USD as default (global standard)
exchange_service = ExchangeRateService(default_base='USD')

File: frontend/src/lib/exchangeRate.ts
/**
 * Exchange Rate Service
 * Fetches live exchange rates from API
 */

const EXCHANGE_API_BASE = 'https://api.exchangerate-api.com/v4/latest';

interface ExchangeRates {
    [currency: string]: number;
}

let cachedRates: ExchangeRates | null = null;
let lastFetchTime = 0;
const CACHE_DURATION = 3600000; // 1 hour in milliseconds

/**
 * Fetch exchange rates from API with caching
 * @param baseCurrency Base currency (default: GBP)
 * @returns Exchange rates object
 */
export async function getExchangeRates(baseCurrency: string = 'GBP'): Promise<ExchangeRates> {
    const now = Date.now();

    // Return cached rates if still valid
    if (cachedRates && (now - lastFetchTime) < CACHE_DURATION) {
        return cachedRates;
    }

    try {
        const response = await fetch(`${EXCHANGE_API_BASE}/${baseCurrency}`);

        if (!response.ok) {
            throw new Error(`Exchange rate API error: ${response.status}`);
        }

        const data = await response.json();

        // Convert rates to "from foreign currency to base currency"
        // API gives us "1 GBP = X USD", we need "1 USD = Y GBP"
        const rates: ExchangeRates = {};
        for (const [currency, rate] of Object.entries(data.rates as ExchangeRates)) {
            rates[currency] = 1 / rate; // Invert the rate
        }

        cachedRates = rates;
        lastFetchTime = now;

        console.log(`✅ Fetched exchange rates for ${Object.keys(rates).length} currencies`);
        return rates;

    } catch (error) {
        console.error('Failed to fetch exchange rates:', error);

        // Fallback to hardcoded rates if API fails
        return getFallbackRates();
    }
}

/**
 * Convert amount from one currency to another
 * @param amount Amount to convert
 * @param fromCurrency Source currency
 * @param toCurrency Target currency (default: GBP)
 * @returns Converted amount
 */
export async function convertCurrency(
    amount: number,
    fromCurrency: string,
    toCurrency: string = 'GBP'
): Promise<number> {
    if (fromCurrency === toCurrency) {
        return amount;
    }

    const rates = await getExchangeRates(toCurrency);
    const rate = rates[fromCurrency];

    if (!rate) {
        throw new Error(`Exchange rate not available for ${fromCurrency} → ${toCurrency}`);
    }

    return amount * rate;
}

/**
 * Fallback rates if API is unavailable
 */
function getFallbackRates(): ExchangeRates {
    return {
        // African currencies
        'NGN': 0.00052,
        'GHS': 0.053,
        'KES': 0.0062,
        'ZAR': 0.044,
        'EGP': 0.016,
        'TZS': 0.00032,
        'UGX': 0.00021,

        // Major currencies
        'USD': 0.79,
        'EUR': 0.86,
        'JPY': 0.0054,
        'CAD': 0.58,
        'AUD': 0.51,
        'CNY': 0.11,
        'INR': 0.0095,
        'BRL': 0.16,
        'MXN': 0.049,
        'GBP': 1.0,
    };
}
